### DISCUSSION TITLE
It is possible to PutFilesWithRequest?

### DISCUSSION THREAD

**Discussion by colegottdank** [ServiceStack/Discuss#23] [Q&A] [answered]
I am building a PUT endpoint which takes fields as well as files. However, I see there is no 'PostFilesWithRequest' variant for PUT methods. Is there anyway around this? I want to test my fluent validation for files in this endpoint, but am unable due to not being able to upload a file.

**Comment by mythz** [comment]
The vast majority of HTTP File Uploads with additional Request params are done with POST so we don't want to duplicate the API surface area to support an unused method.

You can override and [Simulate alternative HTTP Methods](https://docs.servicestack.net/routing#simulate-alternative-http-methods) in ServiceStack with the `X-Http-Method-Override` HTTP Header which you can configure with your HTTP Client with:

```csharp
var client = new JsonServiceClient(baseUrl) {
    RequestFilter = req => req.Headers[HttpHeaders.XHttpMethodOverride] = HttpMethods.Put
};
```

Alternatively you can use the `SendForm/Async` APIs in the new `JsonApiClient` (.NET 6+) to send [MultipartFormDataContent](https://docs.servicestack.net/locode/files#using-httpclient-multipartformdatacontent) requests, with something like:

```csharp
using var uploadRequest = new MultipartFormDataContent()
    .AddParams(new MyRequest {
      //...
    })
    .AddFile(fieldName, fileName, fileStream);

var api = await client.SendFormAsync<MyRequestResponse>(HttpMethods.Put, typeof(MyRequest).ToApiUrl(), uploadRequest);
```

**Reply by colegottdank** [reply]
So, the idea is that you simulate alternative HTTP Methods and then use PostFilesWithRequest? I attempted that and received the same error about an incorrect HttpMethod which is due to the HttpMethod.Post being hardcoded into the PostFilesWithRequest method.

**Comment by mythz** [comment]
Yeah it should override the HTTP Method, can you provide the raw Request/Response headers (i.e. w/o request body). You can use a HTTP sniffer like Fiddler or Wireshark.

**Reply by colegottdank** [reply]
It is failing before a request is sent in the UrlExtensions.ToUrl as the hardcoded HttpMethod.Post is not matching any of my HttpMethods in my route. Do I have to accept POST requests in my route for this to work? If so, that will not work as I only want that route to be a PUT.

**Comment by mythz** [comment, marked as answer]
I've added a change that changes the PostFilesWithRequest APIs to use its [preferred HTTP Method](https://docs.servicestack.net/api-design#all-apis-have-a-preferred-default-method) if one exists, e.g. if you only have a single route with a single verb it will use PUT otherwise you'll need to add the `IPost` marker interface to your Request DTO.

This change is available from v6.1.1+ that's  now [available on MyGet](https://docs.servicestack.net/myget).
